### PR TITLE
[_140] Add OIDC authentication documentation

### PR DESCRIPTION
**Pull request by MartinFlores751**
Documentation.

### CODE REVIEW

File: API.md [hunk 1]
```diff
@@ -18,7 +18,59 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+
+The core advantage of this grant is the flexibility in how it may be applied.
+
+Following is a brief example of authenticating from the command line.
+
+_Command:_
+```bash
+user_and_pass=$(echo -n "<username>:<password>" | base64 -)
+curl -s -X POST -H "Authorization: iRODS $user_and_pass" \
+    http://localhost:<port>/irods-http-api/<version>/authenticate
```
Comment (MartinFlores751) [inline API.md:36]: Can skip the `user_and_pass`, not sure which is clearer.

File: API.md [hunk 2]
```diff
@@ -18,7 +18,59 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+*   Trying [::1]:9000...
+* Connected to localhost (::1) port 9000
+> GET /irods-http-api/0.1.0/authenticate HTTP/1.1
+> Host: localhost:9000
+> User-Agent: curl/8.4.0
+> Accept: */*
+> 
+< HTTP/1.1 302 Found
+< Server: irods_http_api/0.1.0 [d6164b7aa421265e0f9f9d5f7eb69bb67d8e4a3c]
+< Location: http://oidc.example.org/realms/example/protocol/openid-connect/auth?redirect_uri=http%3A%2F%2F127.0.0.1%3A9000%2Firods-http-api%2F0.1.0%2Fauthenticate&response_type=code&state=748fc079-4617-40e0-b4a7-07331494f7c4&scope=openid&client_id=irods_http_api
```
Comment (MartinFlores751) [inline API.md:64]: Ellipsis the long line?

**Review by MartinFlores751 — COMMENTED** [review]
Minor comments, no action unless others interested in them.

File: API.md [hunk 3]
```diff
@@ -18,7 +18,59 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+Authentication is done in the browser.
+
+_Command:_
+```bash
+curl -s http://localhost:<port>/irods-http-api/<version>/authenticate -v
+```
+
+After running the previous command, you should see output similar to the following example output.
+
+_Example output:_
```
Comment (korydraughn) [inline API.md:53]: This can be removed.

File: API.md (hunk 2 above)
Comment (korydraughn) [reply API.md:64]: Yes please.

**Review by korydraughn — COMMENTED** [review]
We're close.

File: API.md (hunk 1 above)
Comment (MartinFlores751) [reply API.md]: Done.

File: API.md (hunk 2 above)
Comment (MartinFlores751) [reply API.md]: Done.

File: API.md [hunk 4]
```diff
@@ -18,7 +18,59 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+- Resource Owner Password Credentials Grant
+- Authorization Code Grant
+
+#### Resource Owner Password Credentials Grant
+
+The core advantage of this grant is the flexibility in how it may be applied.
+
+Following is a brief example of authenticating from the command line.
+
+_Command:_
```
[the remark this answers is not in the document]
Comment (MartinFlores751) [reply API.md]: Done.

File: API.md [hunk 5]
```diff
@@ -18,7 +18,59 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+```bash
+user_and_pass=$(echo -n "<username>:<password>" | base64 -)
+curl -s -X POST -H "Authorization: iRODS $user_and_pass" \
```
[the remark this answers is not in the document]
Comment (MartinFlores751) [reply API.md]: Done.

File: API.md [hunk 6]
```diff
@@ -18,7 +18,59 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+    http://localhost:<port>/irods-http-api/<version>/authenticate
+```
+
+_Returns:_ A string representing a bearer token that can be used to carry out operations as the authenticated user.
```
[the remark this answers is not in the document]
Comment (MartinFlores751) [reply API.md]: Done.

File: API.md [hunk 7]
```diff
@@ -18,7 +18,59 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+
+#### Authorization Code Grant
+
+Using this grant requires a bit more work to extract the token.
+Authentication is done in the browser.
+
+_Command:_
```
[the remark this answers is not in the document]
Comment (MartinFlores751) [reply API.md]: Done.

File: API.md [hunk 8]
```diff
@@ -18,7 +18,59 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+```bash
+curl -s http://localhost:<port>/irods-http-api/<version>/authenticate -v
```
[the remark this answers is not in the document]
Comment (MartinFlores751) [reply API.md]: Done.

File: API.md [hunk 9]
```diff
@@ -18,7 +18,59 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+```
+
+After running the previous command, you should see output similar to the following example output.
```
[the remark this answers is not in the document]
Comment (MartinFlores751) [reply API.md]: Done.

File: API.md (hunk 3 above)
Comment (MartinFlores751) [reply API.md]: Done.

File: API.md [hunk 10]
```diff
@@ -18,7 +18,58 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+
+##### Response
+A string representing a bearer token that can be used to carry out operations as the authenticated user.
+
+#### Authorization Code Grant
+
+Using this grant requires a bit more work to extract the token.
+Authentication is done in the browser.
+
+##### Request
```
Comment (korydraughn) [inline API.md:47]: We don't have a `##### Response` section for this and it's not clear where it should be added.

Maybe this header should become `##### Request / Response`. What do you think?

Comment (korydraughn) [reply API.md:47]: Sure. Let's go with your suggestion.

Comment (MartinFlores751) [reply API.md:47]: Done.

File: API.md [hunk 11]
```diff
@@ -18,7 +18,59 @@ _Returns:_ A string representing a bearer token that can be used to carry out op
...
+```bash
+username_and_password=$(echo -n "<username>:<password>" | base64 -)
+curl -X POST -H "Authorization: iRODS $username_and_password" \
+    http://localhost:<port>/irods-http-api/<version>/authenticate
+```
+
+##### Response
+A string representing a bearer token that can be used to carry out operations as the authenticated user.
```
Comment (korydraughn) [inline API.md:40]: `used to carry out operations` -> `used to execute operations`

**Comment by korydraughn** [comment]
Just one more tweak and I think we're good to go.

File: API.md (hunk 11 above)
Comment (MartinFlores751) [reply API.md]: Done.

**Review by korydraughn — APPROVED** [review]
Pound it.

**Comment by MartinFlores751** [comment]
Pounded.

**Comment by korydraughn** [comment]
Excellent! Merging.

**Pull request merged**